README.md
<div align="center">

<img alt="Icon" width="128" height="128" align="center" src=".github/icon.png"/>

# Loose RGB

**Normalizes RGB color data types to a certain format**

[![npm version](https://badgen.net/npm/v/loose-rgb?icon=npm)](https://www.npmjs.com/package/loose-rgb)
[![check status](https://badgen.net/github/checks/evelynhathaway/loose-rgb/main?icon=github)](https://github.com/evelynhathaway/loose-rgb/actions)
[![minified + gzip bundle size](https://badgen.net/bundlephobia/minzip/loose-rgb)](https://bundlephobia.com/result?p=loose-rgb)
[![license: MIT](https://badgen.net/badge/license/MIT/blue)](/LICENSE)

</div>

## Description

Working with different color libraries gets cumbersome when they use different RGB value formats. Now you can [`normalize`](#normalize) the outputs and [`convert`](#convert) your inputs.

## Features

- Normalize any RGB format to `{r, g, b, [a]}`
- Convert `{r, g, b, [a]}` to any RGB format
- RGBA support
- Hex support

## Installation

```bash
npm install loose-rgb
```

## Usage

```js
import looseRGB from "loose-rgb";

// Convert any type to `{r, g, b, [a]}`
looseRGB.normalize({red: 1, green: 2, blue: 3}) // ? {r: 1, g: 2, b: 3}
looseRGB.normalize("FFBB00AA") // ? {r: 255, g: 187, b: 0, a: 0.66796875}

// Convert `{r, g, b, [a]}` to any type
looseRGB.convert.toCSSValue({r: 255, g: 187, b: 0, a: 0.66796875}) // ? rgba(255, 187, 0, 0.66796875)
looseRGB.convert.toUpperKeys({r: 1, g: 2, b: 3}) // ? {R: 1, G: 2, B: 3}

// Composing the above methods
import {convert, normalize} from "loose-rgb";

const anyToHex = (rgba) => convert.toHex(normalize(rgba));
anyToHex([1, 2, 3, 0.4]) // ? #01020366
```

### Normalize

`looseRGB.normalize(color: RGBUnkown)`

Input any color value as `color` (see [`convert`](#convert)).

### Convert

`looseRGB.convert[to{ColorType}](color: RGBLowerKeys)`

Input an `{r, g, b, [a]}` color to get a reformatted value. See [supported formats](#supported-formats) for methods.

#### Supported Formats

- **`.toLowerKeys()`**
	- `{r: 1, g: 2, b: 3}`
	- `{r: 1, g: 2, b: 3, a: 0.4}`
- **`.toUpperKeys()`**
	- `{R: 1, G: 2, B: 3}`
	- `{R: 1, G: 2, B: 3, A: 0.4}`
- **`.toFullKeys()`**
	- `{red: 1, green: 2, blue: 3}`
	- `{red: 1, green: 2, blue: 3, alpha: 0.4}`
- **`.toArray()`**
	- `[1, 2, 3]`
	- `[1, 2, 3, 0.4]`
- **`.toCSSValue()` —** Supports with and without spacing between commas, leading zero on decimal
	- `"rgb(0, 213, 34)"`
	- `"rgba(0, 213, 34,.12)"`
- **`.toHex()` —** Supports with and without pound, case insensitive, with shorthand
	- `"#fff"`
	- `"#000000"`
	- `"#FFBB00AA"`
	- `"123"`
	- `"FFFF"`

## License

Copyright Evelyn Hathaway, [MIT License](/LICENSE)


CHANGELOG.md
# Changelog

## v1.0.7

**[`v1.0.6...v1.0.7`](https://github.com/evelynhathaway/loose-rgb/compare/v1.0.6...v1.0.7)** - **2021-06-09**

### 📦 Package

- **dev:** Bump @babel/preset-env from 7.14.2 to 7.14.4 [`3646f2b`](https://github.com/evelynhathaway/loose-rgb/commit/3646f2b)
- **dev:** Bump @typescript-eslint/eslint-plugin [`31bd4bc`](https://github.com/evelynhathaway/loose-rgb/commit/31bd4bc)
- **dev:** Bump @typescript-eslint/eslint-plugin [`2030f4a`](https://github.com/evelynhathaway/loose-rgb/commit/2030f4a)
- **dev:** Bump @typescript-eslint/parser from 4.25.0 to 4.26.0 [`897b7e9`](https://github.com/evelynhathaway/loose-rgb/commit/897b7e9)
- **dev:** Bump @typescript-eslint/parser from 4.26.0 to 4.26.1 [`b7f8bfc`](https://github.com/evelynhathaway/loose-rgb/commit/b7f8bfc)
- **dev:** Bump conventional-changelog-evelyn from 1.2.3 to 1.3.0 [`358777c`](https://github.com/evelynhathaway/loose-rgb/commit/358777c)
- **dev:** Bump eslint from 7.26.0 to 7.27.0 [`998a30b`](https://github.com/evelynhathaway/loose-rgb/commit/998a30b)
- **dev:** Bump eslint from 7.27.0 to 7.28.0 [`62143ee`](https://github.com/evelynhathaway/loose-rgb/commit/62143ee)
- **dev:** Bump eslint-plugin-evelyn from 5.0.4 to 5.0.5 [`640ac4b`](https://github.com/evelynhathaway/loose-rgb/commit/640ac4b)
- **dev:** Bump eslint-plugin-evelyn from 5.0.5 to 6.1.0 [`0bfb69a`](https://github.com/evelynhathaway/loose-rgb/commit/0bfb69a)
- **dev:** Bump eslint-plugin-evelyn from 6.1.0 to 6.2.0 [`a97aa7a`](https://github.com/evelynhathaway/loose-rgb/commit/a97aa7a)
- **dev:** Bump eslint-plugin-evelyn from 6.2.0 to 6.2.1 [`98a1a4e`](https://github.com/evelynhathaway/loose-rgb/commit/98a1a4e)
- **dev:** Bump eslint-plugin-import from 2.23.2 to 2.23.3 [`c635f52`](https://github.com/evelynhathaway/loose-rgb/commit/c635f52)
- **dev:** Bump eslint-plugin-import from 2.23.3 to 2.23.4 [`928f242`](https://github.com/evelynhathaway/loose-rgb/commit/928f242)
- **dev:** Bump eslint-plugin-regexp from 0.10.0 to 0.11.0 [`c735a7b`](https://github.com/evelynhathaway/loose-rgb/commit/c735a7b)
- **dev:** Bump jest from 26.6.3 to 27.0.1 [`4dd2e86`](https://github.com/evelynhathaway/loose-rgb/commit/4dd2e86)
- **dev:** Bump jest from 27.0.1 to 27.0.3 [`a39c5b2`](https://github.com/evelynhathaway/loose-rgb/commit/a39c5b2)
- **dev:** Bump jest from 27.0.3 to 27.0.4 [`9cefcd1`](https://github.com/evelynhathaway/loose-rgb/commit/9cefcd1)
- **dev:** Bump ts-jest from 27.0.1 to 27.0.2 [`f744635`](https://github.com/evelynhathaway/loose-rgb/commit/f744635)
- **dev:** Bump ts-jest from 27.0.2 to 27.0.3 [`ec3d0ae`](https://github.com/evelynhathaway/loose-rgb/commit/ec3d0ae)
- **dev:** npm update dev deps [`92c8346`](https://github.com/evelynhathaway/loose-rgb/commit/92c8346)
- **dev:** npm update, linting [`f676ae1`](https://github.com/evelynhathaway/loose-rgb/commit/f676ae1)
- **prod:** Bump ts-essentials from 7.0.1 to 7.0.2 [`080e91f`](https://github.com/evelynhathaway/loose-rgb/commit/080e91f)

## v1.0.6

**[`v1.0.5...v1.0.6`](https://github.com/evelynhathaway/loose-rgb/compare/v1.0.5...v1.0.6)** - **2021-05-18**

### 📦 Package

- **dev:** npm update [`660a2f3`](https://github.com/evelynhathaway/loose-rgb/commit/660a2f3)

## v1.0.5

**[`v1.0.4...v1.0.5`](https://github.com/evelynhathaway/loose-rgb/compare/v1.0.4...v1.0.5)** - **2021-05-18**

### 📦 Package

- **npm:** Bump @babel/cli from 7.13.16 to 7.14.3 [`7942f3d`](https://github.com/evelynhathaway/loose-rgb/commit/7942f3d)
- **npm:** Bump @babel/core from 7.14.2 to 7.14.3 [`2ab1bca`](https://github.com/evelynhathaway/loose-rgb/commit/2ab1bca)
- **npm:** Bump @typescript-eslint/eslint-plugin [`f4d7ba0`](https://github.com/evelynhathaway/loose-rgb/commit/f4d7ba0)
- **npm:** Bump @typescript-eslint/parser from 4.23.0 to 4.24.0 [`4a00a80`](https://github.com/evelynhathaway/loose-rgb/commit/4a00a80)
- **npm:** Bump conventional-changelog-evelyn from 1.2.2 to 1.2.3 [`a59ed23`](https://github.com/evelynhathaway/loose-rgb/commit/a59ed23)

## v1.0.4

**[`v1.0.3...v1.0.4`](https://github.com/evelynhathaway/loose-rgb/compare/v1.0.3...v1.0.4)** - **2021-05-17**

### 📄 Documentation

- **readme:** Update badges [`781623e`](https://github.com/evelynhathaway/loose-rgb/commit/781623e)

## v1.0.3

**[`v1.0.2...v1.0.3`](https://github.com/evelynhathaway/loose-rgb/compare/v1.0.2...v1.0.3)** - **2021-05-17**

### 📦 Package

- **npm:** Deep dependency upgrade [`feef08f`](https://github.com/evelynhathaway/loose-rgb/commit/feef08f)
- **npm:** Update npm dependencies, GitHub Actions [`73d2a5f`](https://github.com/evelynhathaway/loose-rgb/commit/73d2a5f)

### 🔧 Configuration

- **Babel:** Remove false browser query [`b839993`](https://github.com/evelynhathaway/loose-rgb/commit/b839993)

### Other

- Git hooks for linting, change ts target for instrumented tests [`3d51fd1`](https://github.com/evelynhathaway/loose-rgb/commit/3d51fd1)

## v1.0.2

**[`v1.0.1...v1.0.2`](https://github.com/evelynhathaway/loose-rgb/compare/v1.0.1...v1.0.2)** - **2020-07-19**


### 🧹 Internal


- ESM support, transpile modules using Babel, ship source maps, release config [`1b697e2`](https://github.com/evelynhathaway/loose-rgb/commit/1b697e2)


src/test.test.ts
import {normalize, convert} from ".";


const testFelicity = normalize({r: 1, g: 2, b: 3}); // ?
const testLupe = normalize({r: 1, g: 2, b: 3, a: 4}); // ?
const testCecelia = normalize({R: 1, G: 2, B: 3}); // ?
const testMakenzie = normalize({R: 1, G: 2, B: 3, A: 4}); // ?
const testMarlin = normalize({red: 1, green: 2, blue: 3}); // ?
const testEduardo = normalize({red: 1, green: 2, blue: 3, alpha: 4}); // ?
const testBarney = normalize([1, 2, 3]); // ?
const testWanda = normalize([1, 2, 3, 4]); // ?
const testLexie = normalize("rgb(0, 213, 34)"); // ?
const testLauren = normalize("rgba(0, 213, 34,12)"); // ?
const testRosa = normalize("#fff"); // ?
const testSofia = normalize("#FFBB00AA"); // ?
const testTatum = normalize("123"); // ?
const testAntonio = normalize("FFFF"); // ?

const testJose = convert.toLowerKeys({r: 1, g: 2, b: 3}); // ?
const testFloy = convert.toUpperKeys({r: 1, g: 2, b: 3}); // ?
const testJett = convert.toFullKeys({r: 1, g: 2, b: 3}); // ?
const testMafalda = convert.toArray({r: 1, g: 2, b: 3}); // ?
const testLexus = convert.toCSSValue({r: 1, g: 2, b: 3}); // ?
const testVerda = convert.toHex({r: 1, g: 2, b: 3}); // ?
const testMichele = convert.toLowerKeys({r: 1, g: 2, b: 3, a: 4}); // ?
const testPascale = convert.toUpperKeys({r: 1, g: 2, b: 3, a: 4}); // ?
const testLuna = convert.toFullKeys({r: 1, g: 2, b: 3, a: 4}); // ?
const testJadon = convert.toArray({r: 1, g: 2, b: 3, a: 4}); // ?
const testAli = convert.toCSSValue({r: 1, g: 2, b: 3, a: 4}); // ?
const testRosalinda = convert.toHex({r: 1, g: 2, b: 3, a: 4}); // ?


src/__tests__/convert.test.ts
import {RGBUnknown} from "../helpers.js";
import {convert} from "../index.js";


interface GenericConvert {
	(color: RGBUnknown): RGBUnknown;
}
describe("convert", () => {
	const convertMethods: Array<keyof typeof convert> = [
		"toLowerKeys",
		"toUpperKeys",
		"toFullKeys",
		"toArray",
		"toCSSValue",
		"toHex",
	];
	const convertEdgeCases = [
		{r: 0, g: 1, b: 2},
		{r: 255, g: 101, b: 16},
		{r: 0, g: 1, b: 2, a: 0},
		{r: 255, g: 101, b: 16, a: 0.95},
	];

	describe.each(convertMethods)("%s", (method) => {
		test.each(convertEdgeCases)("%p", (color) => {
			expect((convert[method] as GenericConvert)(color)).toMatchSnapshot();
		});
	});
});


src/__tests__/normalize.test.ts
import {normalize} from "../index.js";


describe("normalize", () => {
	const normalizeEdgeCases = [
		{r: 1, g: 2, b: 3},
		{r: 1, g: 2, b: 3, a: 0.4},
		{R: 1, G: 2, B: 3},
		{R: 1, G: 2, B: 3, A: 0.4},
		{red: 1, green: 2, blue: 3},
		{red: 1, green: 2, blue: 3, alpha: 0.4},
		[1, 2, 3],
		[1, 2, 3, 0.4],
		"rgb(0, 213, 34)",
		"rgba(0, 213, 34,.12)",
		"#fff",
		"#000000",
		"#FFBB00AA",
		"123",
		"FFFF",
	];

	test.each(normalizeEdgeCases)("%p", (color) => {
		expect(normalize(color)).toMatchSnapshot();
	});

	test("throw error if not recognized", () => {
		expect(() => {
			(normalize as CallableFunction)({r: 1, a: "string"});
		}).toThrowError("The input into `looseRGB.normalize` wasn't a recognized RGB type.");
	});
});


src/convert.ts
import {MarkOptional} from "ts-essentials";
import {
	RGBLowerKeys,
	RGBALowerKeys,
	RGBUpperKeys,
	RGBAUpperKeys,
	RGBFullKeys,
	RGBAFullKeys,
	RGBArray,
	RGBAArray,
	RGBHex,
	RGBAHex,
	RGBCSSValue,
	RGBACSSValue,
} from "./helpers.js";


/*
	Converting
*/
export function toLowerKeys<InputRGBA extends RGBALowerKeys>(input: InputRGBA): RGBALowerKeys;
export function toLowerKeys<InputRGB extends RGBLowerKeys>(input: InputRGB): RGBLowerKeys;
export function toLowerKeys ({r, g, b, a}: MarkOptional<RGBALowerKeys, "a">): MarkOptional<RGBALowerKeys, "a"> {
	return a ? {r, g, b, a} : {r, g, b};
}
export function toUpperKeys<InputRGBA extends RGBALowerKeys>(input: InputRGBA): RGBAUpperKeys;
export function toUpperKeys<InputRGB extends RGBLowerKeys>(input: InputRGB): RGBUpperKeys;
export function toUpperKeys ({r, g, b, a}: MarkOptional<RGBALowerKeys, "a">): MarkOptional<RGBAUpperKeys, "A"> {
	return a ? {R: r, G: g, B: b, A: a} : {R: r, G: g, B: b};
}
export function toFullKeys<InputRGBA extends RGBALowerKeys>(input: InputRGBA): RGBAFullKeys;
export function toFullKeys<InputRGB extends RGBLowerKeys>(input: InputRGB): RGBFullKeys;
export function toFullKeys ({r, g, b, a}: MarkOptional<RGBALowerKeys, "a">): MarkOptional<RGBAFullKeys, "alpha"> {
	return a ? {red: r, green: g, blue: b, alpha: a} : {red: r, green: g, blue: b};
}
export function toArray<InputRGBA extends RGBALowerKeys>(input: InputRGBA): RGBAArray;
export function toArray<InputRGB extends RGBLowerKeys>(input: InputRGB): RGBArray;
export function toArray ({r, g, b, a}: MarkOptional<RGBALowerKeys, "a">): RGBArray | RGBAArray {
	return a ? [r, g, b, a] : [r, g, b];
}
export function toHex<InputRGBA extends RGBALowerKeys>(input: InputRGBA): RGBAHex;
export function toHex<InputRGB extends RGBLowerKeys>(input: InputRGB): RGBHex;
export function toHex ({r, g, b, a: aFloat}: MarkOptional<RGBALowerKeys, "a">): RGBHex | RGBAHex {
	const aInt = aFloat && Math.ceil(aFloat * 256) - 1;
	const [rHex, gHex, bHex, aHex] = [r, g, b, aInt].map((number) => {
		if (number === undefined) return "";
		const hex = number.toString(16);
		return hex.length === 1 ? `0${hex}` : hex;
	});
	return `#${rHex}${gHex}${bHex}${aHex}`;
}
export function toCSSValue<InputRGBA extends RGBALowerKeys>(input: InputRGBA): RGBACSSValue;
export function toCSSValue<InputRGB extends RGBLowerKeys>(input: InputRGB): RGBCSSValue;
export function toCSSValue ({r, g, b, a}: MarkOptional<RGBALowerKeys, "a">): RGBCSSValue | RGBACSSValue {
	return a ? `rgba(${r}, ${g}, ${b}, ${a})` : `rgb(${r}, ${g}, ${b})`;
}


src/normalize.ts
import {
	RGB,
	RGBLowerKeys,
	RGBA,
	RGBALowerKeys,
	isRGBLowerKeys,
	isRGBUpperKeys,
	isRGBFullKeys,
	isRGBArray,
	isRGBCSSValue,
	regexRGBCSSValue,
	parseDec,
	isRGBACSSValue,
	regexRGBACSSValue,
	isRGBHex,
	regexRGBHex,
	parseHex,
	isRGBAHex,
	regexRGBAHex,
	isRGBHexShort,
	regexRGBHexShort,
	isRGBAHexShort,
	regexRGBAHexShort,
	RGBUnknown,
} from "./helpers.js";
import {convert} from "./index.js";


/*
	Normalizing
*/
export function normalize<InputRGBA extends RGBA>(input: InputRGBA): RGBALowerKeys;
export function normalize<InputRGB extends RGB>(input: InputRGB): RGBLowerKeys;
export function normalize<InputRGBA extends string>(input: InputRGBA): RGBLowerKeys | RGBALowerKeys;
export function normalize (input: RGBUnknown): RGBLowerKeys | RGBALowerKeys;
export function normalize (input: RGBUnknown): RGBLowerKeys | RGBALowerKeys {
	if (isRGBLowerKeys(input)) {
		const {r, g, b, a} = input;
		return convert.toLowerKeys({r, g, b, a});
	}
	if (isRGBUpperKeys(input)) {
		const {R, G, B, A} = input;
		return convert.toLowerKeys({r: R, g: G, b: B, a: A});
	}
	if (isRGBFullKeys(input)) {
		const {red, green, blue, alpha} = input;
		return convert.toLowerKeys({r: red, g: green, b: blue, a: alpha});
	}
	if (isRGBArray(input)) {
		const [r, g, b, a] = input;
		return convert.toLowerKeys({r, g, b, a});
	}
	if (isRGBCSSValue(input)) {
		const [, r, g, b] = regexRGBCSSValue.exec(input)!.map(parseDec);
		return convert.toLowerKeys({r, g, b});
	}
	if (isRGBACSSValue(input)) {
		const [, r, g, b, a] = regexRGBACSSValue.exec(input)!.map(parseDec);
		return convert.toLowerKeys({r, g, b, a});
	}
	if (isRGBHex(input)) {
		const [, r, g, b] = regexRGBHex.exec(input)!.map(parseHex);
		return convert.toLowerKeys({r, g, b});
	}
	if (isRGBAHex(input)) {
		const [, r, g, b, a] = regexRGBAHex.exec(input)!.map(parseHex);
		return convert.toLowerKeys({r, g, b, a});
	}
	if (isRGBHexShort(input)) {
		const [, r, g, b] = regexRGBHexShort.exec(input)!.map((number, index) => parseHex(number.repeat(2), index));
		return convert.toLowerKeys({r, g, b});
	}
	if (isRGBAHexShort(input)) {
		const [, r, g, b, a] = regexRGBAHexShort.exec(input)!.map((number, index) => parseHex(number.repeat(2), index));
		return convert.toLowerKeys({r, g, b, a});
	}

	throw new Error("The input into `looseRGB.normalize` wasn't a recognized RGB type.");
}


src/helpers.ts
import type {MarkOptional} from "ts-essentials";


/*
	Types
*/
export interface RGBLowerKeys {
	r: number;
	g: number;
	b: number;
}
export interface RGBALowerKeys extends RGBLowerKeys {
	a: number;
}

export interface RGBUpperKeys {
	R: number;
	G: number;
	B: number;
}
export interface RGBAUpperKeys extends RGBUpperKeys {
	A: number;
}

export interface RGBFullKeys {
	red: number;
	green: number;
	blue: number;
}
export interface RGBAFullKeys extends RGBFullKeys {
	alpha: number;
}

export type RGBArray = [number, number, number] | Array<number>;
export type RGBAArray = [number, number, number, number] | Array<number>;

export type RGBCSSValue = string;
export type RGBACSSValue = string;

export type RGBHex = string;
export type RGBAHex = string;

export type RGB = RGBLowerKeys | RGBUpperKeys | RGBFullKeys | RGBArray;
export type RGBA = RGBALowerKeys | RGBAUpperKeys | RGBAFullKeys | RGBAArray;
export type RGBX = RGB | RGBA;
export type RGBUnknown = RGB | RGBA | RGBCSSValue | RGBACSSValue | RGBHex | RGBAHex;

export type RGBTypeName = "LowerKeys" | "UpperKeys" | "FullKeys" | "Array" | "Hex" | "CSSValue"


/*
	Helpers
*/
// eslint-disable-next-line @typescript-eslint/ban-types
export const includesKeys = (source: Array<string>, target: Object): boolean => (
	source.every((key) => Object.keys(target).includes(key))
);
export const parseDec = (number: string, index: number): number => (
	// Parse float for alpha
	index === 4 ? Number.parseFloat(number) : Number.parseInt(number, 10)
);
export const parseHex = (number: string, index: number): number => {
	const numberInt = Number.parseInt(number, 16);
	// Convert to float for alpha
	return index === 4 ? (numberInt + 1) / 256 : numberInt;
};


/*
	Regular Expressions
*/
export const regCSSNumber = "\\s*(\\d+)\\s*";
export const regCSSFloatNumber = "\\s*(\\d?\\.?\\d+)\\s*";
export const regexRGBCSSValue = new RegExp(`^rgb\\(${regCSSNumber},${regCSSNumber},${regCSSNumber}\\)$`, "i");
export const regexRGBACSSValue = new RegExp(`^rgba\\(${regCSSNumber},${regCSSNumber},${regCSSNumber},${regCSSFloatNumber}\\)$`, "i");

export const regHexNumber = "([\\dA-F]{2})";
export const regHexShortNumber = "([\\dA-F])";
export const regexRGBHexShort = new RegExp(`^#?${regHexShortNumber}${regHexShortNumber}${regHexShortNumber}$`, "i");
export const regexRGBAHexShort = new RegExp(`^#?${regHexShortNumber}${regHexShortNumber}${regHexShortNumber}${regHexShortNumber}$`, "i");
export const regexRGBHex = new RegExp(`^#?${regHexNumber}${regHexNumber}${regHexNumber}$`, "i");
export const regexRGBAHex = new RegExp(`^#?${regHexNumber}${regHexNumber}${regHexNumber}${regHexNumber}$`, "i");


/*
	Type Guards
*/
export const isRGBLowerKeys = (input: RGBUnknown): input is MarkOptional<RGBALowerKeys, "a"> => (
	input instanceof Object && includesKeys(["r", "g", "b"], input)
);
export const isRGBUpperKeys = (input: RGBUnknown): input is MarkOptional<RGBAUpperKeys, "A"> => (
	input instanceof Object && includesKeys(["R", "G", "B"], input)
);
export const isRGBFullKeys = (input: RGBUnknown): input is MarkOptional<RGBAFullKeys, "alpha"> => (
	input instanceof Object && includesKeys(["red", "green", "blue"], input)
);
export const isRGBArray = (input: RGBUnknown): input is RGBArray | RGBAArray => (
	Array.isArray(input) && input.length >= 3
);
export const isRGBCSSValue = (input: RGBUnknown): input is RGBCSSValue => (
	typeof input === "string" && regexRGBCSSValue.test(input)
);
export const isRGBACSSValue = (input: RGBUnknown): input is RGBACSSValue => (
	typeof input === "string" && regexRGBACSSValue.test(input)
);
export const isRGBHex = (input: RGBUnknown): input is RGBHex => (
	typeof input === "string" && regexRGBHex.test(input)
);
export const isRGBAHex = (input: RGBUnknown): input is RGBAHex => (
	typeof input === "string" && regexRGBAHex.test(input)
);
export const isRGBHexShort = (input: RGBUnknown): input is RGBHex => (
	typeof input === "string" && regexRGBHexShort.test(input)
);
export const isRGBAHexShort = (input: RGBUnknown): input is RGBAHex => (
	typeof input === "string" && regexRGBAHexShort.test(input)
);
